plugins/preserve-tombs.cpp

```diff
@@ -0,0 +1,269 @@
+#include "PluginManager.h"
+#include "MiscUtils.h"
+
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <algorithm>
+#include <utility>
+#include <cstdint>
+
+#include "modules/Units.h"
+#include "modules/Buildings.h"
+#include "modules/Persistence.h"
+#include "modules/EventManager.h"
+#include "modules/World.h"
+
+#include "df/world.h"
+#include "df/unit.h"
+#include "df/building.h"
+#include "df/building_civzonest.h"
+
+using namespace DFHack;
+using namespace df::enums;
+
+
+// <BOILERPLATE>
+DFHACK_PLUGIN("preserve-tombs");
+DFHACK_PLUGIN_IS_ENABLED(is_enabled);
+
+REQUIRE_GLOBAL(world);
+
+
+static const std::string CONFIG_KEY = std::string(plugin_name) + "/config";
+static PersistentDataItem config;
+
+static int32_t cycle_timestamp;
+static int32_t cycle_freq;
+
+enum ConfigValues {
+    CONFIG_IS_ENABLED = 0,
+    CONFIG_CYCLES = 1
+};
+
+static std::unordered_map<int32_t, int32_t> tomb_assignments;
+
+namespace DFHack {
+    DBG_DECLARE(preservetombs, config, DebugCategory::LINFO);
+}
+
+
+static int get_config_val(PersistentDataItem &c, int index) {
+    if (!c.isValid())
+        return -1;
+    return c.ival(index);
+}
+static bool get_config_bool(PersistentDataItem &c, int index) {
+    return get_config_val(c, index) == 1;
+}
+static void set_config_val(PersistentDataItem &c, int index, int value) {
+    if (c.isValid())
+        c.ival(index) = value;
+}
+static void set_config_bool(PersistentDataItem &c, int index, bool value) {
+    set_config_val(c, index, value ? 1 : 0);
+}
+
+static bool assign_to_tomb(int32_t unit_id, int32_t building_id);
+static void update_tomb_assignments(color_ostream& out);
+void onUnitDeath(color_ostream& out, void* ptr);
+static command_result do_command(color_ostream& out, std::vector<std::string>& params);
+
+DFhackCExport command_result plugin_init(color_ostream &out, std::vector <PluginCommand> &commands) {
+    commands.push_back(PluginCommand(
+        plugin_name,
+        "Preserves tomb assignments to units when they die.",
+        do_command));
+    return CR_OK;
+}
+
+static command_result do_command(color_ostream& out, std::vector<std::string>& params) {
+    if (params.size() != 1 || params[0] != "status") {
+        out.print("%s wrong usage", plugin_name);
+        return CR_WRONG_USAGE;
+    }
+    else {
+        out.print("%s is currently %s\n", plugin_name, is_enabled ? "enabled" : "disabled");
+        if (is_enabled) {
+            out.print("tracked tomb assignments:\n");
+            std::for_each(tomb_assignments.begin(), tomb_assignments.end(), [&out](const auto& p){
+                auto& [unit_id, building_id] = p;
+                out.print("unit %d -> building %d\n", unit_id, building_id);
+            });
+        }
+        return CR_OK;
+    }
+}
+
+// event listener
+EventManager::EventHandler assign_tomb_handler(onUnitDeath, 0);
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
+    tomb_assignments.clear();
```

Comment:
this will clear tomb assignments if the plugin is loaded and running and then the player runs `enable preserve-tombs`, which is probably not what you want

Reply:
I've adjusted `plugin_enable` to do `tomb_assignments.clear()` when going from enabled to disabled, and `update_tomb_assignments(out)` vice versa.